### PR TITLE
implement selectdeparser and related deparser like groupby, limit, ...

### PR DESCRIPTION
**Pull request by binkertpat**

### CODE REVIEW

**Review by Alletkla — CHANGES_REQUESTED** [review]

File: src/main/java/sqltoregex/settings/regexgenerator/SpellingMistake.java [hunk 1]
```diff
@@ -30,9 +30,14 @@ public String generateRegExFor(String str){
...
         StringBuilder alternativeWritingStyles = new StringBuilder();
         alternativeWritingStyles.append(isCapturingGroup ? '(' : "(?:");
-        alternativeWritingStyles.append(str).append('|');
+        alternativeWritingStyles.append(str.replace("\\(", "\\\\(").replace("\\)", "\\\\)")).append('|');
         for(int i = 0; i<str.length(); i++){
-            alternativeWritingStyles.append(str, 0, i).append(str, i+1, str.length()).append("|");
+            String first = str.substring(0, i);
+            String second = str.substring(i+1);
+            String concat = first.concat(second);
+            concat = concat.replace("\\(", "\\\\(").replace("\\)", "\\\\)");
```
[the remark this answers is not in the document]
Comment (binkertpat) [reply src/main/java/sqltoregex/settings/regexgenerator/SpellingMistake.java:38]: fixed

File: src/main/java/sqltoregex/settings/regexgenerator/SpellingMistake.java [hunk 2]
```diff
@@ -30,9 +30,14 @@ public String generateRegExFor(String str){
         }
         StringBuilder alternativeWritingStyles = new StringBuilder();
         alternativeWritingStyles.append(isCapturingGroup ? '(' : "(?:");
-        alternativeWritingStyles.append(str).append('|');
+        alternativeWritingStyles.append(str.replace("\\(", "\\\\(").replace("\\)", "\\\\)")).append('|');
```
[the remark this answers is not in the document]
Comment (binkertpat) [reply src/main/java/sqltoregex/settings/regexgenerator/SpellingMistake.java:33]: fixed

File: src/main/java/sqltoregex/settings/SettingsMapBuilder.java [hunk 3]
```diff
@@ -15,13 +16,15 @@ class SettingsMapBuilder {
     private static final String UNSUPPORTED_BUILD_WITH = "Unsupported build with:";
     private static final String STRING_SYNONYM_DELIMITER = ";";
     private final Set<OrderRotation> orderRotations;
+    private final Set<ExpressionRotation> expressionRotations;
```
[the remark this answers is not in the document]
Comment (binkertpat) [reply src/main/java/sqltoregex/settings/SettingsMapBuilder.java]: done

File: src/main/java/sqltoregex/deparser/OrderByDeParserForRegEx.java [hunk 4]
```diff
@@ -3,33 +3,108 @@
...
+        List<String> orderByElementsAsStrings = new LinkedList<>();
+        for(OrderByElement orderByElement : orderByElementList){
+            orderByElementsAsStrings.add(deParseElementForOrderRotation(orderByElement));
+        }
+        buffer.append(orderRotation.generateRegExFor(orderByElementsAsStrings));
+    }
+
+    private String handleAscDesc(OrderByElement orderByElement){
+        StringBuilder temp = new StringBuilder();
+        if (!orderByElement.isAsc()) {
```
[the remark this answers is not in the document]
Comment (binkertpat) [reply src/main/java/sqltoregex/deparser/OrderByDeParserForRegEx.java:103]: done

File: src/main/java/sqltoregex/deparser/OrderByDeParserForRegEx.java [hunk 5]
```diff
@@ -3,33 +3,108 @@
...
+    private static final String DELIMITER_FOR_ORDERROTATION_WITHOUT_SPELLINGMISTAKE = "##########";
+    public static final String ORDER = "ORDER";
+    public static final String SIBLINGS = "SIBLINGS";
+    public static final String BY = "BY";
+    public static final String DESC = "DESC";
+    public static final String ASC = "ASC";
+    public static final String NULLS_LAST = "NULLS LAST";
+    public static final String NULLS_FIRST = "NULLS FIRST";
+    private ExpressionVisitor regExExpressionVisitor;
+    private final boolean isKeywordSpellingMistake;
```
[the remark this answers is not in the document]
Comment (binkertpat) [reply src/main/java/sqltoregex/deparser/OrderByDeParserForRegEx.java]: done

File: src/main/java/sqltoregex/deparser/GroupByDeParserForRegEx.java [hunk 6]
```diff
@@ -1,10 +1,58 @@
...
+import sqltoregex.settings.regexgenerator.ExpressionRotation;
+import sqltoregex.settings.regexgenerator.SpellingMistake;
+
+import java.util.*;
 
 public class GroupByDeParserForRegEx extends GroupByDeParser {
-    public GroupByDeParserForRegEx(ExpressionVisitor expressionVisitor, StringBuilder buffer) {
+    private static final String REQUIRED_WHITE_SPACE = "\\s+";
+    private static final String OPTIONAL_WHITE_SPACE = "\\s*";
+    private final boolean isKeywordSpellingMistake;
```
[the remark this answers is not in the document]
Comment (binkertpat) [reply src/main/java/sqltoregex/deparser/GroupByDeParserForRegEx.java]: done

File: src/test/java/sqltoregex/deparser/LimitDeParserForRegExTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+class LimitDeParserForRegExTest {
+    StringBuilder buffer = new StringBuilder();
+    StatementDeParser statementDeParser = new StatementDeParserForRegEx(new ExpressionDeParserForRegEx(new SettingsManager()), buffer);
+
+    LimitDeParserForRegExTest() throws XPathExpressionException, ParserConfigurationException, IOException, SAXException {
+    }
+
+    boolean checkAgainstRegEx(String regex, String toChecked) {
+        Pattern pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
+        Matcher matcher = pattern.matcher(toChecked);
```
[the remark this answers is not in the document]
Comment (binkertpat) [reply src/test/java/sqltoregex/deparser/LimitDeParserForRegExTest.java:28]: done

File: src/main/java/sqltoregex/deparser/StatementDeParserForRegEx.java [hunk 8]
```diff
@@ -13,8 +24,28 @@ public StatementDeParserForRegEx(ExpressionDeParser expressionDeParser, SelectDe
...
 
+
+    @Override
+    public void visit(Select select) {
+        selectDeParserForRegEx.setBuffer(buffer);
+        expressionDeParserForRegEx.setSelectVisitor(selectDeParserForRegEx);
+        expressionDeParserForRegEx.setBuffer(buffer);
+        selectDeParserForRegEx.setExpressionVisitor(expressionDeParserForRegEx);
+        if (select.getWithItemsList() != null && !select.getWithItemsList().isEmpty()) {
+            buffer.append("WITH ");
```
Comment (Alletkla) [inline src/main/java/sqltoregex/deparser/StatementDeParserForRegEx.java:41]: require whitespace instead of normal whitespace character

**Review by Alletkla — CHANGES_REQUESTED** [review]

**Review by Alletkla — APPROVED** [review]

**Pull request merged**